pkg/exemplars/tsdb.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package exemplars
+
+import (
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/thanos-io/thanos/pkg/exemplars/exemplarspb"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+)
+
+// TSDB implements exemplarspb.Exemplars that allows to fetch exemplars from a TSDB instance.
+type TSDB struct {
+	db        storage.ExemplarQueryable
+	extLabels labels.Labels
+}
+
+// NewTSDB creates new exemplars.TSDB.
+func NewTSDB(db storage.ExemplarQueryable, extLabels labels.Labels) *TSDB {
+	return &TSDB{
+		db:        db,
+		extLabels: extLabels,
+	}
+}
+
+// Exemplars returns all specified exemplars from a TSDB instance.
+func (t *TSDB) Exemplars(r *exemplarspb.ExemplarsRequest, s exemplarspb.Exemplars_ExemplarsServer) error {
+	eq, err := t.db.ExemplarQuerier(s.Context())
+	if err != nil {
+		return err
+	}
+
+	expr, err := parser.ParseExpr(r.Query)
+	if err != nil {
+		return err
+	}
+
+	matchers := parser.ExtractSelectors(expr)
+
+	exemplars, err := eq.Select(r.Start, r.End, matchers...)
+	if err != nil {
+		return err
+	}
```

Comment:
Line 36 - line 46 is the same for all tenants so maybe we can extract this part out the function?

Reply:
The code changed a bit around this part so resolving this for now.

Reply:
not resolved I think.

Reply:
Ah, I misunderstood what you were trying to say. I got it now, let's continue at https://github.com/thanos-io/thanos/pull/4292#discussion_r646047501